Adds Publisher Option To MakeBinding Extension

Twigz opened the pull request:
## What it Does
Adds new extensions to the BindingAdditions to allow for passing in a publisher rather than an action.

## Notes
After seeing the use case of the ViewStore with no actions, I thought this would be a nice addition to condense the creation of bindings who's `set` sent a value to a publisher. In practice, rather than specifying a case path, you can now simply specify a publisher like so:

```swift
makeBinding(viewStateKeyPath: \.searchText, publisher: searchTextPublisher)
```

This is useful for either `ViewStore`s that don't have any actions, or for view stores that want to have `Binding` that doesn't use a publicized action.

I would've used `some Subject`, instead of one for Passthrough and one for CurrentValue,  but that's not out yet.

Pearapps commented:
This _seems_ reasonable, but my one concern / can we verify that you can write tests for bindings made with this approach?

Twigz commented:
@Pearapps So there is an existing test for `testToggleShowsPhotoCount` in the example. I updated the `showsPhotoCount` Binding from 
`makeBinding(viewStateKeyPath: \.showsPhotoCount, actionCasePath: /Action.toggleShowsPhotoCount)`
to
`makeBinding(viewStateKeyPath: \.showsPhotoCount, publisher: showsPhotosCountPublisher)`
and ran the test, which passed.

At the very least this indicates this is *as* testable as what we currently have. Is that sufficient?

Pearapps commented:
> @Pearapps So there is an existing test for `testToggleShowsPhotoCount` in the example. I updated the `showsPhotoCount` Binding from `makeBinding(viewStateKeyPath: \.showsPhotoCount, actionCasePath: /Action.toggleShowsPhotoCount)` to `makeBinding(viewStateKeyPath: \.showsPhotoCount, publisher: showsPhotosCountPublisher)` and ran the test, which passed.
> 
> At the very least this indicates this is _as_ testable as what we currently have. Is that sufficient?

It would still rely on the action, though - which wouldn't be there if you wrote the ViewStore using the new publisher based binding, though, right?

mliberatore reviewed (approved):
Approving, pending resolution on Ken’s comments

Twigz commented:
> > @Pearapps So there is an existing test for `testToggleShowsPhotoCount` in the example. I updated the `showsPhotoCount` Binding from `makeBinding(viewStateKeyPath: \.showsPhotoCount, actionCasePath: /Action.toggleShowsPhotoCount)` to `makeBinding(viewStateKeyPath: \.showsPhotoCount, publisher: showsPhotosCountPublisher)` and ran the test, which passed.
> > At the very least this indicates this is _as_ testable as what we currently have. Is that sufficient?
> 
> It would still rely on the action, though - which wouldn't be there if you wrote the ViewStore using the new publisher based binding, though, right?

@Pearapps 

So I updated the test which was using the action like so: `vs.send(.toggleShowsPhotoCount(true))` to instead just set the wrapped value of the binding `vs.showsPhotoCount.wrappedValue = true`.

Is that what you were looking for as far as testability goes?

Pearapps commented:
> Is that what you were looking for as far as testability goes?

Yes! 

We likely should have an example using this approach in a test so people know how to test it and we can add that in a follow up or just another PR in the future?
